<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="example">

    <material name="white">
        <color rgba="1 1 1 1"/>
    </material>

    <link name="link_0">
        <visual>
            <geometry>
                <mesh filename = "package://fun4/meshes/link_0.stl"/>
            </geometry>
            <material name="white"/>
        </visual>
    </link>

    <link name="link_1">
        <visual>
            <geometry>
                <mesh filename = "package://fun4/meshes/link_1.stl"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <material name="white"/>
        </visual>
    </link>

    <joint name="joint_1" type="revolute">
        <axis rpy="0 0 0" xyz="0 0 1"/>
        <limit 
            effort="1000.0" 
            lower="-3.14159265359" 
            upper="3.14159265359" 
            velocity="1000.0"
            />
        <parent link="link_0"/>
        <child link="link_1"/>
        <origin xyz="0 0 0.2"/>
    </joint>

    <link name="link_2">
        <visual>
            <geometry>
                <mesh filename = "package://fun4/meshes/link_2.stl"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <material name="white"/>
        </visual>
    </link>

    <joint name="joint_2" type="revolute">
        <axis rpy="0 0 0" xyz="0 0 1"/>
        <limit 
            effort="1000.0" 
            lower="-3.14159265359" 
            upper="3.14159265359" 
            velocity="1000.0"
            />    
        <parent link="link_1"/>
        <child link="link_2"/>
        <origin rpy="-1.57075 -1.57075 0" xyz="0 -0.12 0"/>
    </joint>

    <link name="link_3">
        <visual>
            <geometry>
                <mesh filename = "package://fun4/meshes/link_3.stl"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <material name="white"/>
        </visual>
    </link>

    <joint name="joint_3" type="revolute">
        <axis rpy="0 0 0" xyz="0 0 1"/>
        <limit 
            effort="1000.0" 
            lower="-3.14159265359" 
            upper="3.14159265359" 
            velocity="1000.0"
            />    
        <parent link="link_2"/>
        <child link="link_3"/>
        <origin rpy="0 0 0" xyz="0.25 0 0.1"/>
    </joint>

    <link name="end_effector">
        <visual>
            <geometry>
                <mesh filename = "package://fun4/meshes/end_effector.stl"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <material name="white"/>
        </visual>
    </link>

    <joint name="link_3_to_e" type="fixed">
        <parent link="link_3"/>
        <child link="end_effector"/>
        <origin rpy="1.57075 0 1.57075" xyz="0.28 0 0"/>
    </joint>    

</robot>
